feat: 사용자 정보 조회 API 구현

devjaeyoon opened the pull request:
### ✨ 이슈 번호

- #10 

### 📌 설명

사용자 정보 조회 API를 구현하여 작성한 Pull Request입니다.

### 📃 작업 사항

- [x]  `userId`를 Path Parameter로부터 추출한다.
- [x]  사용자 ID에 해당하는 데이터가 DB에 존재하는지 확인한다.
- [x]  사용자 정보가 존재할 경우:
    - [x]  응답 데이터에 `userId`, `isAdDetect`, `isReturnChannel`, `createdAt` , `interestChannels`포함한다.
    - [x]  `200 OK` 로 응답을 반환한다.
- [x]  사용자 정보가 존재하지 않을 경우:
    - [x]  `404` 상태와 "사용자를 찾을 수 없습니다." 응답을 반환한다.
- [x]  서버 오류 발생 시:
    - [x]  `500` 상태와 "서버에서 요청을 처리할 수 없습니다." 응답을 반환한다.

### 💡 참고 사항

 • controllers/userController.js
API 요청을 처리하는 역할입니다. 클라이언트로부터 userId를 받아서, 서비스를 호출해 사용자 정보를 조회하고, 적절한 HTTP 응답을 보냅니다.
 • middlewares/errorHandler.js
서버에서 발생하는 예외(오류)를 한 곳에서 처리해줍니다. 오류가 발생하면 500 상태 코드와 함께 친절한 메시지를 반환합니다.
 • routes/users/[userId].js
사용자 관련 API 경로를 정의합니다. /:userId 경로로 들어오는 요청을 받아서 검증 미들웨어와 컨트롤러로 연결해줍니다.
 • services/userService.js
데이터베이스와 직접 통신하는 부분입니다. userId로 사용자를 조회하고, DB에서 받아온 데이터를 JavaScript 스타일의 카멜 케이스로 변환해 반환합니다.
 • utils/caseConverter.js
DB에서 온 스네이크 케이스(snake_case) 데이터를 카멜 케이스(camelCase)로 바꿔주는 유틸 함수입니다. JavaScript 코드 내에서 변수명 일관성을 유지할 수 있게 도와줍니다.
 • validators/userValidator.js
API 요청에서 전달받은 userId가 숫자인지 검증합니다. 잘못된 형식이면 400 에러를 반환해 서버가 이상한 요청을 받지 않도록 방어합니다.

![image](https://github.com/user-attachments/assets/64a5ede0-2ed6-4601-8421-fe353455ab6c)

![image](https://github.com/user-attachments/assets/a2fbbd67-6e73-4d21-902f-e7dce4489af3)

### 💭 리뷰 사항 반영

- [x] 타입이 숫자인지 검증하는 로직에서 숫자로 변환하는 과정을 진행하지 않았던 오류 수정
- [x] 템플릿 리터럴을 써야하는 문자열을 큰 따옴표로 감쌌던 문제 수정

### ✅ Pull Request 체크 사항

- [x] 가장 최신 브랜치를 pull 하였습니다.
- [x] base 브랜치명을 확인하였습니다.
- [x] 코드 컨벤션을 모두 확인하였습니다.
- [x] 브랜치명을 확인하였습니다.

devjaeyoon committed:
feat: toCamelCase 유틸 함수 추가하여 응답 데이터 카멜 케이스 변환

Supabase에서 가져온 snake_case 응답 데이터를
JS/Express에서 일관성 있게 사용할 수 있도록 camelCase로 변환
중첩된 객체와 배열도 재귀적으로 처리함

devjaeyoon committed:
feat: 사용자 ID 기반 사용자 정보 조회 서비스 로직 구현

Supabase에서 사용자 정보와 관심 채널 정보를 조회하는 서비스 로직 구현
snake_case를 camelCase로 변환하여 일관된 응답 데이터 제공
존재하지 않는 사용자에 대한 null 처리 포함

devjaeyoon committed:
feat: 사용자 ID로 사용자 정보를 조회하는 컨트롤러 구현

요청된 userId를 기반으로 사용자 정보를 조회하고 응답을 반환하는 컨트롤러 구현
존재하지 않는 경우 404 상태 코드와 에러 메시지 반환 처리 포함

devjaeyoon committed:
feat: 사용자 ID 유효성 검증 미들웨어 추가

요청 파라미터로 전달된 userId가 숫자인지 확인하는 미들웨어 구현
잘못된 형식일 경우 400 상태 코드와 함께 에러 메시지 반환

devjaeyoon committed:
feat: 에러 처리 미들웨어 errorHandler 구현

서버 내부 오류 발생 시 500 상태 코드와 에러 메시지를 반환하는
기본 에러 핸들러 미들웨어 추가
Express 기본 시그니처에 맞게 _req 인자 포함

devjaeyoon committed:
feat: 사용자 조회 API 라우터 [userId].js 추가

userId 파라미터를 통해 사용자를 조회하는 GET API 라우팅 설정
userValidator로 파라미터 유효성 검사 후 userController로 요청 처리

devjaeyoon committed:
refactor: 서버 구조 개선 및 라우팅 미들웨어 적용

- 기존 단일 라우트 방식 제거
- /users/:userId 라우트를 routes/users/[userId].js로 분리
- errorHandler 미들웨어 추가
- express.json() 적용으로 body 파싱 처리
- .env의 PORT 값 적용

Eun0713 reviewed (dismissed):
LGTM👍

faber-h commented on src/middlewares/errorHandler.js:
```diff
@@ -0,0 +1,8 @@
+export const errorHandler = (err, _req, res) => {
```
`req` 앞에만 `_`를 붙이신 이유가 궁금합니다~

devjaeyoon replied:
`_req`를 사용한 이유는 express의 에러 핸들링 미들웨어 시그니처에 맞추기 위해 `res` 앞에 `req`를 작성해야 합니다. 여기서 실제 사용하지 않은 변수를 포함하면 ESLint가 no-unused-vars 에러를 출력하는데 이 에러를 해결하고자 `req` 앞에 `_`를 붙여 실제 `req`를 사용하지 않으면서 ESLint의 에러를 해결하기 위해 작성했습니다.

faber-h commented on src/validators/userValidator.js:
```diff
@@ -0,0 +1,12 @@
+export const validateUserId = (req, res, next) => {
+  const { userId } = req.params;
+
+  if (Number.isNaN(userId)) {
```
params로 가져온 userId는 문자열이라서 Number.isNaN(userId)를 하면 무조건 false가 나와서 검사가 안될 것 같습니다~
문자열 변환하고 조건문을 쓰는건 어떨까요?

devjaeyoon replied:
맞습니다. `Number()`를 이용하여 문자열을 숫자로 변환하는 작업을 선행하고 만약, 문자열이라면 `NaN`이 될 테니 그 값을 `Number.isNaN()`으로 확인하는 절차로 진행하겠습니다. 좋은 리뷰 감사드립니다.

d0-neii commented on src/server.js:
```diff
@@ -1,21 +1,20 @@
-  console.log("Server running on port 3000");
+const PORT = process.env.PORT || 3000;
+app.listen(PORT, () => {
+  console.log("🚀 Server running on port ${PORT}");
```
이 부분은 템플릿 리터럴(`)로 감싸야할 것 같습니다!😊

devjaeyoon replied:
짚어주셔서 감사합니다. 수정하겠습니다!

devjaeyoon committed:
fix: 문자열 userId가 NaN으로 제대로 검증되지 않는 문제 수정

- 기존에는 'Number.isNaN(userId)'로 검사했지만, 문자열이 들어올 경우 'isNaN('123')'
- false로 검증이 제대로 동작하지 않음
- 'Number(userId)'로 변환한 후 'Number.isNaN(...)'을 적용해
- 정확하게 숫자 여부를 판별하도록 수정

devjaeyoon committed:
fix: 서버 포트 출력 로그에서 문자열 템플릿 문법 수정

- 기존에 포트 출력 시 일반 문자열로 작성되어 PORT 변수가 출력되지 않았음
- 백틱을 사용해 템플릿 리터럴로 수정하여 실제 포트 번호가 출력되도록 수정

faber-h reviewed (approved):
LGTM🔥

Eun0713 reviewed (approved):
LGTM🌟
